Sundials 5.0 Interface [sundials-interface]

gardner48 opened the pull request:
New interface for Sundials 5+. I've created this as a draft PR to keep from merging it prematurely since the interfaces use new functions in Sundials that have not been released yet.

The PR includes two approaches for attaching an MFEM linear solver to Sundials. The first uses a `SundialsLinearSolver` class (like the current interface) and the second depends on new `TimeDependentOperator` methods specific to Sundials (similar to methods used with MFEM's integrators or the PETSc interface). I've included both approaches so we can discuss which method is the preferred. 

The PR also adds support for ARKode's IMEX methods (by adding a second `TimeDependentOperator` to the `ODESolver` class) and setting up the ODE in linearly implicit form (ARKode's native mass matrix support).

Because the internal memory structures for Sundials packages are no longer exposed, it's necessary to call `Init` for CVODE/ARKode or `SetOperator` for KINSOL before calling any methods to set package options.

acfisher commented on linalg/ode.hpp:
```diff
@@ -24,9 +24,10 @@ class ODESolver
 protected:
    /// Pointer to the associated TimeDependentOperator.
    TimeDependentOperator *f;  // f(.,t) : R^n --> R^n
+   TimeDependentOperator *f2; // f(.,t) : R^n --> R^n
```
Why do we need 2 of these now?

gardner48 replied:
This was added for using IMEX methods with ARKode.

acfisher replied:
That makes sense.

acfisher commented on linalg/operator.hpp:
```diff
@@ -261,6 +261,68 @@ class TimeDependentOperator : public Operator
       return const_cast<Operator &>(dynamic_cast<const Operator &>(*this));
    }
 
+   /** Setup the ODE linear system A(y,t) = (I - gamma J) or A = (M - gamma J).
```
I'm kinda surprised to see these methods baked into Operator rather than in the ODE classes.  Do these really belong here?

gardner48 replied:
This is one of the two approaches I mentioned for attaching an MFEM linear solver to Sundials. This idea here was to mirror the `ImplicitSolve` function used by MFEM's integrators. Sundials integrators need `ImplicitSetup` in addition to `ImplicitSovle` because they will try to reuse the same linear system across Newton iterations and time steps.

acfisher replied:
It does look like the TimeDependentOperator already has the solve concept in it so this is just going along with that.  I hadn't see that recently and it seems backwards to me, but rolling with the way MFEM is handling the integrators it the best choice.

acfisher replied:
@v-dobrev Julian and I are happy with the new way this is implemented here.  Please have a look to make sure you agree with this approach.  If so we'll review this further.

jandrej commented:
I think the new approach which utilizes the modified `TimeDependentOperator` is very clear to understand. Especially since it benefits from the recent interface changes in Sundials that now expose a generic linear solver and matrix interface.

It also streamlines the approach to incorporate other solvers (from Sundials) and mitigates the effect of having to keep up with the internals, since it just feeds the interface.

jandrej commented:
@gardner48 I'd like to put this in the next branch to get momentum on the PR. There is a merge conflict which should be resolved easily. If you could make the necessary changes I will merge it into next before the weekend tests start.

gardner48 commented:
@jandrej Sounds good, I'll take care of that now. I'll also note that we posted a new beta release on Monday (`sundials-v5.0.0-dev.1`) that includes the new Sundials functions needed in this interface. You can get it from the [development branch of our GitHub repo](https://github.com/LLNL/sundials/tree/develop) or the [Sundials LLNL download page](https://computation.llnl.gov/projects/sundials/sundials-software).

jandrej commented:
> @jandrej Sounds good, I'll take care of that now. I'll also note that we posted a new beta release on Monday (`sundials-v5.0.0-dev.1`) that includes the new Sundials functions needed in this interface. You can get it from the [development branch of our GitHub repo](https://github.com/LLNL/sundials/tree/develop) or the [Sundials LLNL download page](https://computation.llnl.gov/projects/sundials/sundials-software).

Thanks! Merged into next for testing.

jandrej commented on linalg/sundials.hpp:
```diff
@@ -24,385 +24,476 @@
+
+  public:
+
+    /// Construct a serial warpper to SUNDIALS' KINSOL nonlinear solver
```
s/warpper/wrapper

acfisher commented:
Here is our current TODO list on this pull request:

- [ ] SUNDIALS 5.0 Release (David)
- [ ] Update INSTALL and CHANGELOG (Aaron)
- [ ] Remove 't' from the argument list of TimeDependentOperator::ImplicitSetup/MassSetup (Aaron)
- [ ] Swap'x', 'b' args in TimeDependentOperator::ImplicitSolve/MassSolve (Aaron)
- [ ] Add SUN Prefix to Implicit/Mass Setup/Solve (Aaron)
- [ ] Remove the additional 'old' style implementation. (David)
- [ ] Update all the SUNDIALS examples to use the new implementation. (David)
- [ ] Update the help strings of the SUNDIALS examples to reflect the new implicit/explicity methods (David)
- [ ] Move ODESolver::Init(f1,f2) definition down in object hierarchy to AREKStepSolver, or perhaps a new IMEXSolver class. (David)
- [ ] Rework KINSolver initializaition with reinit capability from SUNDIALS. (David)
- [ ] Final doxygen check. (Aaron)

gardner48 commented:
>     * In both classes CVODESolver and ARKStepSolver, the method:
>       virtual void Init(TimeDependentOperator &f_);
>       is replaced by the method:
>       void Init(TimeDependentOperator &f_, double &t, Vector &x);
>       which breaks the base class interface.

This is something that came up when we did the walk-through of the interface and I thought there as a way to work around this change using resize/reinit. I've added a resize function to CVODE and pushed the updates to use `virtual void Init(TimeDependentOperator &f_);` and removed `void Init(TimeDependentOperator &f_, double &t, Vector &x);`. To try this out you'll need to build Sundials from the `feature/mfem-updates` branch from our development repo.
 
>     * Do we want to rename/rework the SetLinearSolver() methods?

We'd talked about attaching the interface functions to an MFEM linear solver by default so a user would not need to call this function and it could be removed. Instead a function call would be needed to use one of Sundials' linear solvers. I was going to call this `UseSundialsLinearSolver` or `UseSunidalsGMRES`.

acfisher commented:
@gardner48 Excellent, those were a couple of items on Veselin's list that I wasn't sure about.  It looks like we are getting close to being able to review this.

v-dobrev commented on INSTALL:
```diff
@@ -470,6 +470,7 @@ The specific libraries and their options are:
 - SUNDIALS (optional), used when MFEM_USE_SUNDIALS = YES.
   Beginning with MFEM v3.3, SUNDIALS v2.7.0 is supported.
   Beginning with MFEM v3.3.2, SUNDIALS v3.0.0 is also supported.
+  Beginning with MFEM v4.0.2, only SUNDIALS v5.0.0+ is supported.
```
Shouldn't this be v4.1?

v-dobrev commented:
Running the sample run (in `examples/sundials`)
```console
./ex9 -m ../../data/periodic-square.mesh -p 1 -r 2 -s 8 -dt 0.005 -tf 9
```
I can see some oscillations in the solution in GLVis. It looks like the maximum time step set with `arkode->SetMaxStep(dt)` is not respected, e.g. the final step prints:
```console
time step: 1800, time: 9
ARKStep:
num steps:                 1041
num exp rhs evals:         6564
num imp rhs evals:         0
num lin setups:            0
num nonlin sol iters:      0
num nonlin conv fail:      0
num steps attempted:       1104
num acc limited steps:     1042
num exp limited stepfails: 0
num error test fails:      63
initial dt:                1.71613e-07
last dt:                   0.00839669
current dt:                0.00839669
current t:                 9.00282
```
Maybe I'm not using the right branch -- I use the `develop` branch (from the internal LLNL repo) since I do not see the branch `feature/mfem-updates` mentioned above.

v-dobrev commented on linalg/sundials.hpp:
```diff
@@ -23,384 +23,433 @@
+   void UseMFEMLinearSolver();
+
+   /// Attach SUNDIALS linear solver to CVODE.
+   void UseSundialsLinearSolver();
```
In the above two methods (`UseMFEMLinearSolver` and `UseSundialsLinearSolver`), where is the solver that being set? It will be nice to clarify this in the doxygen documentation.

v-dobrev commented on linalg/sundials.hpp:
```diff
@@ -23,384 +23,433 @@
-   /// Set a custom Jacobian system solver for implicit methods.
-   void SetLinearSolver(SundialsODELinearSolver &ls_spec);
+   /// Attach a SUNDIALS linear solver to ARKode.
+   void UseSundialsLinearSolver();
```
Similar to the `CVODESolver`, it will be nice to clarify which `MFEM`/`SUNDIALS` solvers will be used when the methods `UseMFEMLinearSolver`/`UseSundialsLinearSolver` are called.

v-dobrev commented on linalg/sundials.hpp:
```diff
@@ -23,384 +23,433 @@
+   /// Attach a SUNDIALS mass matrix linear solver to ARKode.
+   /** @param[in] tdep    An integer flag indicating if the mass matrix is time
+                          dependent (1) or time independent (0) */
+   void UseSundialsMassLinearSolver(int tdep);
```
Similar to above, which `MFEM`/`SUNDIALS` solvers will be used when `UseMFEMMassLinearSolver`/`UseSundialsMassLinearSolver` are called?

v-dobrev commented:
This may be due to using the wrong branch but also I see issues with the following `ex10` run:
```console
./ex10 -m ../../data/beam-quad.mesh -r 2 -o 2 -s 12 -dt 0.15 -vs 10
```
For example, the `ΔTE` (change in total energy) seems to blow up:
```console
step 150, t = 22.5, EE = 3.54124, KE = 0.0386086, ΔTE = 3.54747
CVODE:
num steps:            152
num rhs evals:        155
num lin setups:       10
num nonlin sol iters: 152
num nonlin conv fail: 0
num error test fails: 0
last order:           1
current order:        1
initial dt:           5.14839e-06
last dt:              0.15
current dt:           0.15
current t:            22.5515
```
In `master` with SUNDIALS 3.2.1, the numbers are:
```console
step 150, t = 22.5, EE = 0.0129975, KE = 0.0157191, ΔTE = -0.00366434
CVODE:
  num steps: 152, num evals: 156, num lin setups: 10, num nonlin sol iters: 152
  last order: 1, next order: 1, last dt: 0.15, next dt: 0.15
```

gardner48 commented:
> I can see some oscillations in the solution in GLVis. It looks like the maximum time step set with `arkode->SetMaxStep(dt)` is not respected, e.g. the final step prints:

When `ARKStepReInit()` is called in `Step()` it is resetting the max step size to the default rather than retaining the user set value. This is a change in how `ReInit()` worked in earlier versions ARKode. I've put up a PR to restore the old behavior of retaining the max step size when calling `ReInit()`.

> Maybe I'm not using the right branch -- I use the `develop` branch (from the internal LLNL repo) since I do not see the branch `feature/mfem-updates` mentioned above.

I ended up working around the changes made in the `feature/mfem-updates` branch so the `develop` branch in the LLNL repo a good branch to work with.

gardner48 commented:
> This may be due to using the wrong branch but also I see issues with the following `ex10` run:

I hadn't updated the solve linear function in ex10 and ex10p correctly. These are fixed now.

gardner48 commented:
> I've put up a PR to restore the old behavior of retaining the max step size when calling `ReInit()`.

The PR to restore the old `ReInit()` behavior has been merged into the develop branch.

acfisher commented:
It looks to me like this is ready to pull into master.  @v-dobrev, any objections?

acfisher commented:
@v-dobrev

v-dobrev commented on linalg/sundials.cpp:
```diff
@@ -659,331 +668,556 @@ void ARKODESolver::Step(Vector &x, double &t, double &dt)
+
+void ARKStepSolver::SetIMEXTableNum(int etable_num, int itable_num)
+{
+   flag = ARKStepSetTableNum(sundials_mem, itable_num, itable_num);
```
Shouldn't one of the parameters use `etable_num`? (I noticed this because my editor shows `etable_num` as unused.)

gardner48 replied:
Yes, the second input to `ARKStepSteTableNum` should be `etable_num`. I'll fix this in a sec.

v-dobrev commented on examples/sundials/ex10.cpp:
```diff
@@ -4,16 +4,14 @@
-//    ex10 -m ../../data/beam-quad.mesh -r 2 -o 2 -s  5 -dt 0.15 -vs 10
-//    ex10 -m ../../data/beam-tri.mesh  -r 2 -o 2 -s  7 -dt 0.3  -vs 5
-//    ex10 -m ../../data/beam-hex.mesh  -r 1 -o 2 -s  5 -dt 0.2  -vs 5
+//    ex10 -m ../../data/beam-quad.mesh -r 2 -o 2 -s 12 -dt 0.1  -vs 10
```
I see that `dt` had to be decreased here -- is this due to changed behavior/defaults in CVode? I'd expect values that used to work with previous versions to continue to work with the new version of SUNDIALS.

gardner48 replied:
@v-dobrev This change is due to some logic in the new linear solver interface that was bypassed by the old linear solver interface when supplying an external linear solver. If the initial residual of the linear system in the Newton solve is sufficiently small then the linear solve is skipped to save the cost of the solve. In this test, skipping the linear solve results in a smaller maximum step size. I suspect this would not be an issue with tighter tolerances but we could also add a user-callable set routine to disable this check.

acfisher replied:
Veselin is sounds like a minor behavior change that is reasonable.  Do you have a problem with this?  We need to get this pulled in ASAP for the upcoming XSDK release.

v-dobrev replied:
@acfisher, after some offline discussions with @gardner48 we found an appropriate solution and I already pushed the related updates. In other words, this issue has been resolved.

acfisher replied:
Excellent.  Let's go ahead and get it pulled into master then.  Can you hit the approve button?
